### PR TITLE
[TACHYON-590] Allow wildcard in shell commands

### PR DESCRIPTION
**Pull request by tianyin**

### CODE REVIEW

File: common/src/main/java/tachyon/TachyonURI.java [hunk 1]
```diff
@@ -228,6 +231,50 @@ public TachyonURI getParent() {
   }
 
   /**
+   * Get the first n component of the TachyonURI path.
+   * There is no trailing separator as the path will be normalized by normalizePath() 
+   * 
+   * <pre>
```
Comment (uronce-cc) [inline common/src/main/java/tachyon/TachyonURI.java]: /a/b/c, 0   -> ?

File: common/src/main/java/tachyon/TachyonURI.java [hunk 2]
```diff
@@ -228,6 +231,50 @@ public TachyonURI getParent() {
...
+   * /a/b/c, 1              -> /a
+   * /a/b/c, 2              -> /a/b
+   * /a/b/c, 3              -> /a/b/c
```
Comment (uronce-cc) [inline common/src/main/java/tachyon/TachyonURI.java]: /a/b/c, 4  -> ?

File: common/src/main/java/tachyon/TachyonURI.java [hunk 3]
```diff
@@ -228,6 +231,50 @@ public TachyonURI getParent() {
...
+   * </pre>
+   * 
+   * @return the first n path component, null if the path does not n components
```
Comment (uronce-cc) [inline common/src/main/java/tachyon/TachyonURI.java]: typo. null if the path has less than n components?

File: common/src/main/java/tachyon/TachyonURI.java [hunk 4]
```diff
@@ -228,6 +231,50 @@ public TachyonURI getParent() {
...
+   */
+  public String getPathComponent(int n) {
+    int numComp = getNumPathComponents();
+    if (numComp < n) {
+      return null;
+    } else if (numComp == n) {
+      return mUri.getPath();
+    } else {
+      String path = mUri.getPath();
+      String[] comp = path.split(SEPARATOR);
```
Comment (uronce-cc) [inline common/src/main/java/tachyon/TachyonURI.java]: duplicate `split` with that in `getNumPathComponents`.

File: common/src/main/java/tachyon/TachyonURI.java [hunk 5]
```diff
@@ -228,6 +231,50 @@ public TachyonURI getParent() {
...
+      String[] nComp = new String[n];
```
Comment (uronce-cc) [inline common/src/main/java/tachyon/TachyonURI.java]: we may not need to allocate a new array using `StringUtils.join(Arrays.asList(comp).subList(0, n), SEPARATOR)`.

File: common/src/main/java/tachyon/TachyonURI.java [hunk 6]
```diff
@@ -228,6 +231,50 @@ public TachyonURI getParent() {
...
+      for (int i = 0; i < n; i++) {
```
Comment (uronce-cc) [inline common/src/main/java/tachyon/TachyonURI.java]: format issue: `i ++`.

 Also, `"/a/b/c".split("/")` will return `["", "a", "b", "c"]`, so it should be `for (int i = 1; i <= n; i ++)`?

File: common/src/main/java/tachyon/TachyonURI.java [hunk 7]
```diff
@@ -228,6 +231,50 @@ public TachyonURI getParent() {
...
+      }
+      return StringUtils.join(nComp, SEPARATOR);
+    }
+  }
+  
+  /**
+   * Get the number of path components of the given string
+   * @return
+   */
+  public int getNumPathComponents() {
```
Comment (uronce-cc) [inline common/src/main/java/tachyon/TachyonURI.java]: please give some examples, according to current code, "/" -> 0, "/a" -> 2, but I think you want "/a" -> 1?

Comment (uronce-cc) [inline common/src/main/java/tachyon/TachyonURI.java]: hmm, is it the same with `getDepth()` ?

**Comment by apc999** [comment]
@uronce-cc @calvinjia , please review this PR.

**Comment by apc999** [comment]
Please add a link to the JIRA ticket in this PR's description, to ease tracking activities in the future. 

Thanks

**Comment by calvinjia** [comment]
@tianyin Thanks for the patch, could you please resolve uronce-cc's comments?

**Comment by tianyin** [comment]
@calvinjia @uronce-cc @apc999 Thanks! Just came back from Mexico :-S I'll address them asap!

File: common/src/main/java/tachyon/TachyonURI.java (hunk 1 above)
Comment (tianyin) [inline common/src/main/java/tachyon/TachyonURI.java]: /a/b/c, 0 -> /
added, thanks!

File: common/src/main/java/tachyon/TachyonURI.java (hunk 2 above)
Comment (tianyin) [inline common/src/main/java/tachyon/TachyonURI.java]: /a/b/c, 4 -> null
added, thanks!

File: common/src/main/java/tachyon/TachyonURI.java (hunk 3 above)
Comment (tianyin) [inline common/src/main/java/tachyon/TachyonURI.java]: changed, thanks!

File: common/src/main/java/tachyon/TachyonURI.java (hunk 5 above)
Comment (tianyin) [inline common/src/main/java/tachyon/TachyonURI.java]: cool! Changed to this, thanks!

File: common/src/main/java/tachyon/TachyonURI.java (hunk 4 above)
Comment (tianyin) [inline common/src/main/java/tachyon/TachyonURI.java]: Now I use getDepth() so there's no duplicate. Thanks!

File: common/src/main/java/tachyon/TachyonURI.java (hunk 6 above)
Comment (tianyin) [inline common/src/main/java/tachyon/TachyonURI.java]: yes, this is buggy,,,
I used getDepth() now so there's no this issue :-)

File: common/src/main/java/tachyon/TachyonURI.java (hunk 7 above)
Comment (tianyin) [inline common/src/main/java/tachyon/TachyonURI.java]: Exactly! So I'm using getDepth() now :-)

**Comment by calvinjia** [comment]
@tianyin Instead of adding the additional wildcard logic to each of the methods, would it be possible to do the logic at the beginning and then just change it to that many calls of the original method? For example `copyToLocal(/files/*)` would become just n calls of `copyToLocal(/files/1), copyToLocal(/files/2)...`

File: common/src/main/java/tachyon/TachyonURI.java [hunk 8]
```diff
@@ -228,6 +232,43 @@ public TachyonURI getParent() {
...
+   * /a/b/c, 0              -> /
+   * /a/b/c, 1              -> /a
+   * /a/b/c, 2              -> /a/b
+   * /a/b/c, 3              -> /a/b/c
+   * /a/b/c, 4              -> null
+   * </pre>
+   * 
+   * @return the first n path component, null if the path has less than n components
+   */
+  public String getPathComponent(int n) {
```
Comment (apc999) [inline common/src/main/java/tachyon/TachyonURI.java]: NIT: the current name `getPathComponent` is a bit confusing to me. How about `getLeadingPath`?

File: common/src/main/java/tachyon/TachyonURI.java [hunk 9]
```diff
@@ -228,6 +232,43 @@ public TachyonURI getParent() {
...
+    String path = mUri.getPath();
+    if (n == 0 && path.charAt(0) == '/') { //special case
```
Comment (apc999) [inline common/src/main/java/tachyon/TachyonURI.java]: `'/'` ->  `SEPARATOR` ?

File: shell/src/main/java/tachyon/shell/TFsShell.java [hunk 10]
```diff
@@ -133,17 +138,39 @@ public int copyFromLocal(String[] argv) throws IOException {
...
+    if (srcFiles.size() == 0) {
       System.out.println("Local path " + srcPath + " does not exist.");
       return -1;
     }
-    TachyonFS tachyonClient = createFS(dstPath);
-    int ret = copyPath(src, tachyonClient, dstPath);
-    if (ret == 0) {
-      System.out.println("Copied " + srcPath + " to " + dstPath);
+
+    if (srcPath.contains("*") == false) {
```
Comment (apc999) [inline shell/src/main/java/tachyon/shell/TFsShell.java]: shouldn't "*" be `TachyonURI.SEPARATOR`?

**Comment by apc999** [comment]
IMO, this PR is growing too big to review. I suggest to split this PR into multiple smaller PRs---all targets the same JIRA number.
- The first PR has 
`clients/unshaded/src/main/java/tachyon/client/TachyonFS.java`,
`common/src/main/java/tachyon/TachyonURI.java` and their tests. 
- In the rest PR, each PR address one or two shell command to support wildcard, and add its unittests.

**Comment by tianyin** [comment]
@calvinjia I can do this, but the logic looks a bit unnatural to me, as we would need another wrapper to first check whether the command has a wildcard (if so call method A, else call method B). The current implementation uses one entry point. 

In fact, for each method, the logic does not change at all. I just reuse the original implementation and pass all the test cases. 

Let me know your thoughts.

**Comment by tianyin** [comment]
@apc999 ok. I will do this. Yes, I will change the names for sure. Thank you!

**Comment by calvinjia** [comment]
@tianyin 
I think you can just default to calling method A which then calls method B (the real logic) the appropriate times. You will still only have 1 entry point. You can also have a method that evaluates the wildcard path to the appropriate list of paths to operate on.

**Comment by apc999** [comment]
@tianyin shall we close this PR since new incremental PRs are coming?

**Comment by tianyin** [comment]
sorry for the delay (just came back the bay).

Yes, let's close it! Thanks, @apc999 !

**Pull request closed without merging**